### PR TITLE
Customize test reporting for all runs

### PR DESCRIPTION
**Pull request by jrodbx**
Closes https://github.com/cashapp/paparazzi/issues/1651.  Closes https://github.com/cashapp/paparazzi/issues/1671.

https://github.com/cashapp/paparazzi/pull/1563 introduced the inclusion of the failure delta diff in the Junit test results reporting.  However, there are a few noticeable differences from the default reporting.

1) It's based on Google's screenshot plugin reporting, which runs tests in a separate source set and task.  As a result, it also generates its own test result xmls to incorporate custom properties (variant, flavor, failure diff path, etc.). However, it's not clear how this variant/flavor will be used/helpful.
2) It adds extra column headers like "errors" (vs failures) and "skipped" (canonically, referred to as "ignored" in typical reports).
3) The difference between class-level tabs of "Failed Tests" vs "Tests" have changed a bit.  Instead of "Tests" being a tabular view, it's now similar to "Failed Tests" showing an empty diff and empty stack trace code panel, reducing its value.
4) It avoids the built-in Gradle build parallelism (slight counterpoint: they're internal classes, which isn't ideal)
5) It's only implemented for the verify flow, so record flow will use the default Gradle test reporting, allowing for potential issues with Gradle caching and file outputs/structure.
6) The image diff size is rather large
image

This PR address each of those by:
* Using the Gradle implementation of TestReporter and its related classes instead of AGP's screenshot tool (this fixes 1-4)
* Uses the same TestReporter for all runs (fixes 5)
* Renders the image diff in a collapsible pane (fixes 6)

<table>
<tr><th></th><th>Current</th><th>Proposed</th></tr>

<tr>
<td>Summary: Failed Tests</td>
<td>
Screenshot 2024-11-04 at 4 31 37 PM
</td>
<td>
Screenshot 2024-11-04 at 3 29 01 PM
</td>
</tr>

<tr>
<td>Summary: Packages</td>
<td>
Screenshot 2024-11-04 at 4 32 17 PM
</td>
<td>
Screenshot 2024-11-04 at 3 29 07 PM
</td>
</tr>

<tr>
<td>Summary: Classes</td>
<td>
Screenshot 2024-11-04 at 4 32 20 PM
</td>
<td>
Screenshot 2024-11-04 at 3 29 10 PM
</td>
</tr>

<tr>
<td>Class Level: Failed Tests</td>
<td>
Screenshot 2024-11-04 at 4 32 27 PM
</td>
<td>
Screenshot 2024-11-04 at 3 29 42 PM
</td>
</tr>

<tr>
<td>Class Level: All Tests</td>
<td>
Screenshot 2024-11-04 at 4 32 57 PM
</td>
<td>
Screenshot 2024-11-04 at 3 30 01 PM
</td>
</tr>

</table>

And here's how a "Class Level: All Tests" currently looks with mixed results:
Screenshot 2024-11-04 at 5 00 02 PM

### CODE REVIEW

**Comment by jrodbx** [comment]
Did a quick sanity check with non-Paparazzi test failures and Paparazzi test failures that weren't image diff assertions.  Both look good, so I'm ready to promote this.

### Overall Summary of mixed types of failures
Screenshot 2024-11-05 at 1 39 21 PM

### Class Summary with Non-Paparazzi test failure
Screenshot 2024-11-05 at 1 39 27 PM

### Class Summary with mixed Paparazzi test failures (diff assertion vs test exception)
Screenshot 2024-11-05 at 1 39 37 PM

File: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/reporting/ClassPageRenderer.kt [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  private val diffImages: List<DiffImage> =
+    if (isVerifyRun && failureDir.exists()) {
+      failureDir.listFiles()
+        ?.filter { it.name.startsWith("delta-") }
+        ?.map { diff ->
+          // TODO: read from failure diff metadata file instead of brittle parsing
+          val nameSegments = diff.name.split("_", limit = 3)
+          val testClassPackage = nameSegments[0].replace("delta-", "")
+          val testClass = "$testClassPackage.${nameSegments[1]}"
+          val testMethodWithLabel = nameSegments[2].split(".")[0]
```
Comment (geoff-powell) [inline paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/reporting/ClassPageRenderer.kt:31]: nit: this misses some snapshots. I have a fix for this here: https://github.com/cashapp/paparazzi/pull/1677

**Review by geoff-powell — APPROVED** [review]

File: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/reporting/ClassPageRenderer.kt (hunk 1 above)
Comment (geoff-powell) [reply paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/reporting/ClassPageRenderer.kt:31]: Lets get this in first, then I can update my branch.

Comment (jrodbx) [reply paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/reporting/ClassPageRenderer.kt]: Cherry-picked your [commit](https://github.com/cashapp/paparazzi/pull/1675/commits/aaba36d1a53fb6eae521acfec905edd62ac03eaf) for attribution with a little [cleanup](https://github.com/cashapp/paparazzi/pull/1675/commits/badc30eb17bfdfe2f9494c69cf6b5a64bc86f6ce).

**Pull request merged**